README.md
# fairMLHealth
Tools and tutorials for variation analysis in healthcare machine learning.

## New in Version 1.0.0!
We reorganized the library to make things more intuitive and added some useful capabilities:
* Evaluate measures for regression analysis
* Stack stratified tables by cohort groups
* Apply a single **compare** function to evaluate one or multiple models

## Tool Contents
- ### [Documentation and References](./docs/README.md)
    - [Evaluation of Fairness](./docs/resources/Evaluating_Fairness.md)
    - [Quick Reference of Measures](./docs/resources/Measures_QuickReference.md)
    - [References and Resources](./docs/resources/References_and_Resources.md)
    - [Our Publications](./docs/publications/README.md)

- ### [Examples and Tutorials](examples_and_tutorials/README.md)
    - Tutorials for measuring and analyzing fairness as it applies to machine learning
    - Examples for using the templates and tools

- ### [FairMLHealth](fairmlhealth/README.md)
    - **Measure**:
        - Measure healthcare-specific fairness metrics for classification and regression models.
    - **Compare**:
        - Compare protected attributes and other features of a model or several models for potential biases in outcomes.
    - **Visualize**:
        - Visualize biases across multiple models through helpful reports & customizable highlights.

- ### [Templates](templates/README.md)
    - Quickstart notebooks that serve as skeletons for your model analysis

## Installation <a id="installation_instructions"></a>
Installing with pip:

    python -m pip install fairmlhealth

Installing directly from GitHub:

    python -m pip install git+https://github.com/KenSciResearch/fairMLHealth

Installing from a local copy of the repo:

    pip install <path_to_fairMLHealth_dir>

### Troubleshooting Installation Issues

*Trouble with the Installation?*
     - Step 1: Verify that you're using a compatible version of python.
     - Step 2: If step 1 does not resolve your issue, verify that all required packages are properly installed.
     - Step 3: For some metrics, FairMLHealth relies on AIF360, which has a few known installation gotchas. If you are having trouble with your installation, first check [AIF360's Troubleshooting Tips](https://github.com/Trusted-AI/AIF360#troubleshooting).

If you are not able to resolve your issue through the troubleshooting tips above, please let us know through the [Discussion Board](https://github.com/KenSciResearch/fairMLHealth/discussions) or by submitting an issue using the [Issue Template](./docs/code_contributions/ISSUE_TEMPLATE.md) found in our [Documentation folder](./docs/README.md).

## FairMLHealth Usage
Below are some quickstart examples of our most popular features. More information about these and other examples can be found in our [examples_and_tutorials](./examples_and_tutorials) folder! These specific examples are based on our ToolUsage notebooks, for which we've provided online access in Jupyter's nbviewer via the following links:
* [Tool Usage for Binary Classification](https://nbviewer.jupyter.org/github/KenSciResearch/fairMLHealth/blob/master/examples_and_tutorials/Example-ToolUsage_BinaryClassification.ipynb)
* [Tool Usage for Regression](https://nbviewer.jupyter.org/github/KenSciResearch/fairMLHealth/blob/master/examples_and_tutorials/Example-ToolUsage_Regression.ipynb)

Note that while the examples below use pandas, the library is designed to accept either pandas objects or numpy arrays.
### Example Setup

```python
from fairmlhealth import report, measure
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import BernoulliNB
from sklearn.tree import DecisionTreeClassifier


# First we'll create a semi-randomized dataframe with specific columns for our attributes of interest
rng = np.random.RandomState(506)
N = 240
X = pd.DataFrame({'col1': rng.randint(1, 4, N),
                  'col2': rng.randint(1, 75, N),
                  'col3': rng.randint(0, 2, N),
                  'gender': [0, 1]*int(N/2),
                  'ethnicity': [1, 1, 0, 0]*int(N/4),
                  'other': [1, 0, 1, 0, 1, 1, 0, 1]*int(N/8)
                 })
```

### Generalized Reports - report.py
As in previous versions, the latest version of fairMLHealth has tools to create generalized reports of model bias and performance. *An important update starting in Version 1.0.0 is that all of these features are now contained in the **report.py** module (previously named model_comparison.py).*

#### Comparing Binary Classifications
The primary reporting tool is now the **compare** function, which can be used to generate side-by-side comparisons for any number of models, and for either binary classifcation or for regression problems. Model performance metrics such as accuracy and precision (or MAE and RSquared for regression problems) are also provided to facilitate comparison. Below is an example output comparing the two example models defined above.

A flagging protocol is applied by default to highlight any cells with values that are out of range.  This can be turned off by passing ***flag_oor = False*** to report.compare().

``` python

# Ceate a randomized target variable
y = pd.Series(X['col3'].values + rng.randint(0, 2, N), name='Example_Target').clip(upper=1)

# Third, we'll split the data and use it to train two generic models
splits = train_test_split(X, y, stratify=y, test_size=0.5, random_state=60)
X_train, X_test, y_train, y_test = splits

model_1 = BernoulliNB().fit(X_train, y_train)
model_2 = DecisionTreeClassifier().fit(X_train, y_train)

# Generate the report
fairness_measures = report.compare(X_test, y_test, X_test['gender', model_1)

```
Note that the Equal Odds Ratio has been dropped from the example below. This because the false positive rate is approximately zero for both the entire dataset and for the privileged class, leading to a zero in the denominator of the False Positive Rate Ratio. The result is therefore undefined and cannot be compared in the Equal Odds Ratio.

<img src="./docs/img/tool_examples/bin_oneModel.png"
     alt="single model compare example"
     width="800px"
    />


The **compare** tool can also be used to measure two different models or two different protected attributes. Protected attributes are measured separately and cannot yet be combined together with the **compare** tool, although they can be grouped as cohorts in the stratified tables [as shown below](#cohort).


```python
# Example with multiple models
report.compare(test_data = X_test,
               targets = y_test,
               protected_attr = X_test['gender'],
               models = {'Any Name 1':model_1, 'Model 2':model_2})
```
<img src="./docs/img/tool_examples/bin_multiModel.png"
     alt="two model comparison example"
     width="90%"
     />

```python
# Example with different protected attributes.
# Note that the same model is passed with two different keys to clarify the column names.
report.compare(X_test, y_test,
               [X_test['gender'], X_test['ethnicity']],
               {'Gender':model_1, 'Ethnicity':model_1})
```
<img src="./docs/img/tool_examples/bin_multiAttribute.png"
     alt="multiple attribute comparison example"
     width="90%"
     />

#### Comparing Regressions

Here is an example applying the same function for a regression model. Note that the "fair" range to be used for evaluation of regression metrics does requires judgement on the part of the user. Default ranges have been set to [0.8, 1.2] for ratios, 10% of the available target range for *Mean Prediction Difference*, and 10% of the available MAE range for *MAE Difference*. If the default flags do not meet your needs, they can be turned off by passing ***flag_oor = False*** to report.compare(). More information is available in our [Evaluating Fairness Documentation](./docs/resources/Evaluating_Fairness.md#regression_ranges).

``` python
from sklearn.linear_model import LinearRegression

# Create a randomized target variable. In this case we'll add some correlation with existing variables
y = pd.Series((X['col3']+X['gender']).values + rng.uniform(0, 6, N), name='Example_Continuous_Target')

# Split the data and use it to train a regression model
splits = train_test_split(X, y, test_size=0.5, random_state=42)
X_train, X_test, y_train, y_test = splits
regression_model = LinearRegression().fit(X_train, y_train)

# Generate the report.
# Note that for regression models, the prediction type (pred_type) must be declared as such.
report.compare(X_test, y_test, X_test['gender'], regression_model, pred_type="regression")
```
<img src="./docs/img/tool_examples/reg_oneModel.png"
     alt="regression comparison example"
     width="90%"
     />

``` python
# Display the same report with no flags and no model performance
report.compare(X_test, y_test, X_test['gender'], regression_model, pred_type="regression",
                flag_oor=False, skip_performance=True))
```
<img src="./docs/img/tool_examples/reg_skipPerformance.png"
     alt="regression comparison example with skipped performance"
     width="90%"
     />


### Detailed Analyses - measure.py

FairMLHealth also provides tools for detailed analysis of model variance by way of stratified data, performance, and bias tables. Beyond evaluating fairness, these tools are intended for flexible use in any generic assessment of model bais to evaluate multiple features at once. *An important update starting in Version 1.0.0 is that all of these features are now contained in the **measure.py** module (previously named reports.py).*

All tables display a summary row for "All Features, All Values". This summary can be turned off by passing ***add_overview=False*** to measure.data().

#### Stratified Data Tables

The stratified data table can be used to evaluate data against one or multiple targets. Two methods are available for identifying which features to assess, as shown in the first example below.

```python
# The following two function calls will produce the same output table shown below

# Arguments Option 1: pass full set of data, subsetting with *features* argument
measure.data(X_test, y_test, features=['gender'])

# Arguments Option 2: pass the data subset of interest without using the *features* argument
measure.data(X_test[['gender']], y_test)

```
<img src="./docs/img/tool_examples/data_table.png"
     alt="data table example"
     width="90%"
     />

```python
# Display a similar report for multiple targets, dropping the summary row
measure.data(X=X_test, # used to define rows
             Y=X_test, # used to define columns
             features=['gender', 'col1'], # optional subset of X
             targets=['col2', 'col3'], # optional subset of Y
             add_overview=False # turns off "All Features, All Values" row
             )

```
<img src="./docs/img/tool_examples/data_table_multiTarget_noOverview.png"
     alt="multi-target data table example"
     width="90%"
     />

#### Stratified Performance Tables

The stratified performance table evaluates model performance specific to each feature-value subset. These tables are compatible with both classification and regression models. For classification models with the *predict_proba()* method, additional ROC_AUC and PR_AUC values will be included if possible.

```python
# Binary classification performance table with probabilities included
measure.performance(X_test[['gender']],
                    y_true=y_test,
                    y_pred=model_1.predict(X_test),
                    y_prob=model_1.predict_proba(X_test)[:,1])
```
<img src="./docs/img/tool_examples/bin_performance_probs.png"
     alt="performance table example, binary classification"
     width="90%"
     />

```python
# Regression example
measure.performance(X_test[['gender']],
                    y_true=y_test,
                    y_pred=regression_model.predict(X_test),
                    pred_type="regression")
```
<img src="./docs/img/tool_examples/reg_performance.png"
     alt="performance table example, regression"
     width="90%"
     />

#### Stratified Bias Tables

The stratified bias analysis table apply fairness-related metrics for each feature-value pair. It assumes a given feature-value as the "privileged" group relative to all other possible values for the feature. For example, row **2** in the table below displays measures for **"col1"** with a value of **"2"**. For this row, "2" is considered to be the privileged group, while all other non-null values (namely "1" and "3") are considered unprivileged.

To simplify the table, fairness measures have been reduced to their component parts. For example, the Equal Odds Ratio has been reduced to the True Positive Rate (TPR) Ratio and False Positive Rate (FPR) Ratio.

```python
# Binary classification example
# Note that flag_oor is set to False by default for this feature

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. Please do follow the format according to <https://keepachangelog.com/en/1.0.0/>.

_“Yesterday I was clever, so I wanted to change the world. Today I am wise, so I am changing myself.”_ - Rumi

## [1.0.3] - 2021-11-10

### Documentation Update

- Corrections to some math formatting in documentation that could not be resolved by GitHub pages
- Minor updates to text in tutorials, code comments, and other documentation

## [1.0.2] - 2021-09-13
### Pages Hotfix
- Corrections to some math formatting in documentation that could not be resolved by GitHub pages

## [1.0.1] - 2021-09-02

### Supplemental Example Bug Fixes

- Bug fixed in supplemental example notebook

## [1.0.0] - 2021-09-02

### Version 1.0

- Updated setup to use config (setup.cfg) to to align with pypi recommendations
- Cleaned dependencies and package validation (dropped unused modules)
- Updated tutorial validation as appropriate in line with the above changes

## [0.1.29] - 2021-08-05

### Tutorial Update

- Updated tutorials with current functions
- Bug fixes as discovered in tutorials

## [0.1.28] - 2021-08-05

### Documentation Update

- Updated references and images in main README.md
- Updated ToolUsage notebooks in support of main README.md
- Regression measure descriptions added to docs/references
- Updated templates

## [0.1.27] - 2021-08-05

### Bug Fixes

- Attach messages to validation errors
- Prevent errant requests for classification analysis on regeression models
- Update notebook tests so that error and warning evaluation is centralized, can handle broken url strings

## [0.1.26] - 2021-08-05

### Report Update

- Allow flexible return type for any report.compare result
- Update tests for report.py
- Update method for testing if a model predicts probabilities in FairCompare
- Allow user to skip performance measures when using report.compare

## [0.1.25] - 2021-08-05

### Custom Boundary Bug

- Fix bug that prevented custom boundaries from updating

## [0.1.24] - 2021-08-03

### Undefined Ratios and Fairness Metric Testing

- Added code to return NaN values in cases where a fairness ratio has a zero denominator
- Updates to test__fairnessmetrics to increase robustness.
- Ceiling placed on the version number for AIF360 in setup.py

## [0.1.23] - 2021-08-03

### Correct Column References

- Bug: private names appearing in measure tables
- Updated references to private columns

## [0.1.22] - 2021-08-03

### Report Update

- report.measure_model removed in favor of using single function report.compare (neé compare_models) for all reports

## [0.1.21] - 2021-08-03

### Flag-Cohort Correction

- Updates to flagging function to facilitate use for cohorted tables

## [0.1.20] - 2021-08-03

### Analytical Update

- Bootstrap_significance updated to accept any function that returns a p-value, and supporting functions for kruskal and chisquare were added
- Cohorting enabled for measure.data and measure.performance
- Fixed bugs uncovered during testing

## [0.1.19] - 2021-08-03

### Testing Update

- New tests added
- Testing utilities renamed for clarity
- Fixed bugs uncovered during testing

## [0.1.18] - 2021-07-31

### Regression Measure Flagging

- Updated flagging feature to enable custom boundaries and flagging for regression measures
- Related updates to the __validation module. This includes the addition of class-related validation and a first pass at improved type-hinting throughout the tool.
- Minor changes to text displays.

## [0.1.17] - 2021-07-31

### Library Reorganization

- Major changes to the structure of the library intended to make it more intuitive:
  - fairMLHealth modules were renamed to clarify scope and purpose. This will facilitate adding new features in upcoming PRs:
    - **reports.py** is now named **measure.py**
    - **model_comparison** is now named **report.py**
    - utils.py is now split into stat_utils.py and __utils.py
      - __utils.py contains back-end functions for which validation is assumed to have been run
      - stat_utils.py contains supplemental functions that may be useful in statistical analysis or data munging
  - newly-renamed measure.py has also been reorganized so that functions are in alphabetical order and names are more intuitive. (this should have been a separate PR. apologies in advance).
  - load_mimic_data.py is now "__mimic_data.py"
  - the "tutorials_and_examples" folder has been renamed "examples_and_tutorials" so that it will be easier to find (e.g. seen adjacent to the "docs" folder)

## [0.1.16] - 2021-07-30

### Documentation Update

- Doctstings on public-facing functions given added detail components of the code.
- The "feature_table" function was moved into utils.py so that it can be used publicly.

## [0.1.15] - 2021-06-25

### Cohorting Feature

- Enables regression versions of analytical functions, with basic measures added for the regression version of bias analysis table.
- Adds tutorial, template, and example notebooks for regression features
- Improves management of significant figures as required to enable flagging of regression outputs

## [0.1.14] - 2021-06-25

### Cohorting Feature

- Adds cohorting wrapper which iterates to create separate analysis tables by group

## [0.1.13] - 2021-06-23

### Flagging Updates

- Add flagging functionality for stratified tables (bias analysis table, performance analysis table, data analysis table)
- Update flagging function s.t. it can be called as an argument in the model_comparison or bias analysis table functions

## [0.1.12] - 2021-06-17

### Stratified Table Updates

- data analysis table now accepts y as a data frame, enabling stratified analysis across multiple targets.
- "Overview" column evaluating all features as one unit ("ALL_FEATURES") is now optional via add_overview argument

## [0.1.11] - 2021-06-17

### Validation Updates


fairmlhealth/__utils.py
"""
Back-end functions used throughout the library, many of which assume that inputs
have been validated
"""
from numbers import Number
import numpy as np
import pandas as pd
from typing import Any, Callable, Dict, Tuple
from warnings import warn, catch_warnings, filterwarnings

from . import __preprocessing as prep, __validation as valid
from .__validation import ArrayLike, MatrixLike, ValidationError


def epsilon():
    """ error value used to prevent 'division by 0' errors """
    return np.finfo(np.float64).eps


def format_errwarn(func: Callable[[], Tuple[Any, Dict, Dict]]):
    """ Wraps a function returning some result with dictionaries for errors and
        warnings, then formats those errors and warnings as grouped warnings.
        Used for analytical functions to skip errors (and warnings) while
        providing

    Args:
        func (function): any function returning a tuple of format:
            (result, error_dictionary, warning_dictionary). Note that
            dictionaries should be of form {<column or id>:<message>}

    Returns:
        Any (result): the first member of the tuple returned by func
    """

    def format_info(dict):
        info_dict = {}
        for colname, err_wrn in dict.items():
            _ew = list(set(err_wrn)) if isinstance(err_wrn, list) else [err_wrn]
            for m in _ew:
                m = getattr(m, "message") if "message" in dir(m) else str(m)
                if m in info_dict.keys():
                    info_dict[m].append(colname)
                else:
                    info_dict[m] = [colname]
        info_dict = {ew: list(set(c)) for ew, c in info_dict.items()}
        return info_dict

    def wrapper(*args, **kwargs):
        res, errs, warns = func(*args, **kwargs)
        if any(errs):
            err_dict = format_info(errs)
            for er, cols in err_dict.items():
                warn(f"Error processing column(s) {cols}. {er}\n")
        if any(warns):
            warn_dict = format_info(warns)
            for wr, cols in warn_dict.items():
                warn(f"Possible error in column(s) {cols}. {wr}\n")
        return res

    return wrapper


def iterate_cohorts(func: Callable[[], pd.DataFrame]):
    """ Runs the function for each cohort subset

    Args:
        func (function): the function to iterate

    Returns:
        pd.DataFrame: cohort-iterated version of the output

    """

    def prepend_cohort(df: pd.DataFrame, new_ix: ArrayLike):
        """ Attaches cohort information to the far left of the dataframe, adjusting
            by type. Cohorts are added to the index to enable flagging for
            cohorted summary tables (pd.Styler fails with non-unique indices)

        Returns:
            pd.DataFrame
        """
        idx = df.index.to_frame().rename(columns={0: "__index"})
        # add padding to shift start location if prepending a summary table
        ixpad = 0
        if df.index.names == ["Metric", "Measure"]:
            ixpad = 2
        # Attach new_ix data to the index
        if idx.any().any():
            for l, i in enumerate(new_ix):
                idx.insert(l + ixpad, i[0], i[1])
            if "__index" in idx.columns:
                idx.drop("__index", axis=1, inplace=True)
        else:
            pass  # No data in dataframe
        df.index = pd.MultiIndex.from_frame(idx)
        return df

    def subset(data: pd.DataFrame, idxs: ArrayLike):
        if data is not None:
            return data.loc[
                idxs,
            ]
        else:
            return None

    def add_group_info(errant_list, grp_cols, grp_vals):
        grpname = ""
        for i, c in enumerate(grp_cols):
            if len(grpname) > 0:
                grpname += " & "
            grpname += f"{c} {grp_vals[i]}"
        errant_list.append(grpname)
        return None

    def wrapper(cohort_labels: MatrixLike = None, **kwargs):
        """ Iterates for each cohort subset

        Args:
            cohorts (array-like or dataframe, optional): Groups by which to
                subset the data. Defaults to None.

        Returns:
            pandas DataFrame
        """
        # Run preprocessing to facilitate subsetting
        X = kwargs.get("X", None)
        Y = kwargs.get("Y", None)
        y_true = kwargs.get("y_true", None)
        y_pred = kwargs.get("y_pred", None)
        y_prob = kwargs.get("y_prob", None)
        prtc_attr = kwargs.get("prtc_attr", None)
        X, prtc_attr, y_true, y_pred, y_prob = prep.standard_preprocess(
            X, prtc_attr, y_true, y_pred, y_prob
        )
        #
        if cohort_labels is not None:
            #
            valid.validate_data(
                cohort_labels, name="cohort_labels", expected_len=X.shape[0]
            )
            cohorts = prep.prep_data(cohort_labels)
            #
            cix = cohorts.index
            grp_cols = cohorts.columns.tolist()
            cgrp = cohorts.groupby(grp_cols)
            valid.limit_alert(
                cgrp,
                "permutations of cohorts",
                8,
                issue="This may slow processing time and reduce utility.",
            )
            #
            results = []
            errant_list = []
            minobs = valid.MIN_OBS
            for k in cgrp.groups.keys():
                grp_vals = cgrp.get_group(k)[grp_cols].head(1).values[0]
                # Subset each argument to those observations matching the group
                ixs = cix.astype("int64").isin(cgrp.groups[k])
                # Test subset for sufficent data. cohorts with too few observations
                # will cause terminating errors.
                x = subset(X, ixs)
                if x.shape[0] < minobs:
                    add_group_info(errant_list, grp_cols, grp_vals)
                    continue
                # subset remaining arguments
                y = subset(Y, ixs)
                yt = subset(y_true, ixs)
                yh = subset(y_pred, ixs)
                yp = subset(y_prob, ixs)
                pa = subset(prtc_attr, ixs)
                #
                new_args = ["X", "Y", "y_true", "y_pred", "y_prob", "prtc_attr"]
                sub_args = {k: v for k, v in kwargs.items() if k not in new_args}
                try:
                    df = func(
                        X=x,
                        Y=y,
                        y_true=yt,
                        y_pred=yh,
                        y_prob=yp,
                        prtc_attr=pa,
                        **sub_args,
                    )
                except ValidationError as e:
                    # Skip groups for which subsetting led to insufficient target representation
                    m = getattr(e, "message") if "message" in dir(e) else None
                    if m == "Only one target classification found.":
                        df = pd.DataFrame()
                    else:
                        raise ValidationError(m)
                # Empty dataframes indicate issues with evaluation of the function,
                # likely caused by presence of i.e. only one feature-value pair.
                if len(df) == 0:
                    add_group_info(errant_list, grp_cols, grp_vals)
                ix = [(c, grp_vals[i]) for i, c in enumerate(grp_cols)]
                df = prepend_cohort(df, ix)
                results.append(df)
            # Report issues to the user
            if len(errant_list) == len(cgrp.groups.keys()):
                msg = (
                    "Invalid cohort specification. None of the cohort subsets "
                    + f"could be processed. least {minobs} observations."
                )
                raise ValidationError(msg)
            elif any(errant_list):
                msg = (
                    "Could not evaluate function for group(s): "
                    + "{errant_list}. This is commonly caused when there is "
                    + " too little data or there is only a single "
                    + "feature-value pair is available in a given cohort. "
                    + f"Each cohort must have  {minobs} observations."
                )
                warn(msg)
            else:
                pass
            # Combine results and return
            output = pd.concat(results, axis=0)
            return output
        else:
            return func(**kwargs)

    return wrapper


class FairRanges:
    __diffs = [
        "auc difference",
        "balanced accuracy difference",
        "equal odds difference",
        "fpr diff",
        "tpr diff",
        "ppv diff" "positive predictive parity difference",
        "statistical parity difference",
        "mean prediction difference",
        "mae difference",
        "r2 difference",
    ]
    __ratios = [
        "balanced accuracy ratio",
        "disparate impact ratio",
        "selection ratio",
        "equal odds ratio",
        "fpr ratio",
        "tpr ratio",
        "ppv ratio",
        "mean prediction ratio",
        "mae ratio",
        "r2 ratio",
    ]
    __stats = ["consistency score"]

    def __init__(self):
        pass

    def mad(self, arr: ArrayLike):
        return np.median(np.abs(arr - np.median(arr)))

    def load_fair_ranges(
        self,
        custom_ranges: Dict[str, Tuple[Number, Number]] = None,
        y_true: ArrayLike = None,
        y_pred: ArrayLike = None,
    ):
        """
        Args:
            custom_ranges (dict): a  dict whose keys are present among the measures
                in df and whose values are tuples containing the (lower, upper)
                bounds to the "fair" range. If None, uses default boundaries and
                will skip difference measures for regressions models. Default is None.
            y_true (ArrayLike, optional): Sample targets. Defaults to None.
            y_pred (ArrayLike, optional): Sample target predictions. Defaults to None.
        """
        # Load generic defaults
        bnds = self.default_boundaries()

        # Update with specific regression boundaries if possible
        if y_true is not None and y_pred is not None:
            y = prep.prep_arraylike(y_true, "y")
            yh = prep.prep_arraylike(y_pred, "y")
            aerr = (y - yh).abs()
            # Use np.concatenate to combine values s.t. ignore column names
            all_vals = y.append(yh)
            bnds["mean prediction difference"] = self.__calc_diff_range(all_vals)
            bnds["mae difference"] = self.__calc_diff_range(aerr)
        else:
            bnds.pop("mean prediction difference")
            bnds.pop("mae difference")
        #
        if custom_ranges is not None:
            if valid.is_dictlike(custom_ranges):
                for k, v in custom_ranges.items():
                    bnds[str(k).lower().strip()] = v
            else:
                raise TypeError("custom boundaries must be dict-like object if defined")
        #
        available_measures = self.__diffs + self.__ratios + self.__stats
        valid.validate_fair_boundaries(bnds, available_measures)
        return bnds

    def __calc_diff_range(self, ser: pd.Series):
        s_range = np.max(ser) - np.min(ser)
        s_bnd = 0.1 * s_range
        if s_bnd == 0 or np.isnan(s_bnd):
            raise ValidationError(
                "Error computing fair boundaries. Verify targets and predictions."
            )
        return (-s_bnd, s_bnd)

    def default_boundaries(
        self,
        diff_bnd: Tuple[Number, Number] = (-0.1, 0.1),
        rto_bnd: Tuple[Number, Number] = (0.8, 1.2),
    ):
        default = {}
        for d in self.__diffs:
            default[d] = diff_bnd
        for r in self.__ratios:
            default[r] = rto_bnd
        return default


class Flagger:
    __hex = {
        "magenta": "#d00095",
        "magenta_lt": "#ff05b8",
        "purple": "#947fed",
        "purple_lt": "#c2bae3",
    }

    def __init__(self):
        self.reset()

    def apply_flag(
        self,
        df: pd.DataFrame,
        caption: str = "",
        sig_fig: int = 4,
        as_styler: bool = True,
        boundaries: Dict[str, Tuple[Number, Number]] = None,
    ):
        """ Generates embedded html pandas styler table containing a highlighted
            version of a model comparison dataframe
        Args:
            df (pandas dataframe): report.compare dataframe
            caption (str, optional): Optional caption for table. Defaults to "".
            as_styler (bool, optional): If True, returns a pandas Styler of the
                highlighted table (to which other styles/highlights can be added).
                Otherwise, returns the table as an embedded HTML object.

        Returns:
            Embedded html or pandas.io.formats.style.Styler
        """
        self.reset()
        self.__set_df(df)
        self.__set_labels()
        self.__set_boundaries(boundaries)
        #
        if caption is None:
            caption = "Fairness Measures"

        # bools are treated as a subclass of int, so must test for both
        if not isinstance(sig_fig, int) or isinstance(sig_fig, bool):
            raise ValueError(f"Invalid value of significant figure: {sig_fig}")
        if self.label_type == "index":
            styled = self.df.style.set_caption(caption).apply(self.__colors, axis=1)
        else:
            # pd.Styler doesn't support non-unique indices
            if len(self.df.index.unique()) < len(self.df):
                self.df.reset_index(inplace=True)
            styled = self.df.style.set_caption(caption).apply(self.__colors, axis=0)
        # Styler automatically sets precision to 6 sig figs
        with catch_warnings(record=False):
            filterwarnings("ignore", category=DeprecationWarning)
            styled = styled.set_precision(sig_fig)

        #
        setattr(styled, "fair_ranges", self.boundaries)
        # return pandas styler if requested
        if as_styler:
            return styled
        else:
            return styled.render()

    def reset(self):
        """ Clears the __Flagger settings
        """
        self.boundaries = None
        self.df = None
        self.flag_type = "background-color"
        self.flag_color = self.__hex["purple_lt"]
        self.labels = None
        self.label_type = None

    def __colors(self, vals: pd.Series):
        """ Returns a list containing the color settings for difference
            measures found to be OOR
        """

        def is_oor(name, val):
            low = self.boundaries[name.lower()][0]
            high = self.boundaries[name.lower()][1]
            return bool(not low < val < high and not np.isnan(val))

        #
        clr = f"{self.flag_type}:{self.flag_color}"
        if self.label_type == "index":
            name = vals.name[1].lower().strip()
        else:
            name = vals.name.lower().strip()
        #
        if name not in self.boundaries.keys():
            return ["" for v in vals]
        else:
            clr = f"{self.flag_type}:{self.flag_color}"
            return [clr if is_oor(name, v) else "" for v in vals]

    def __set_boundaries(
        self, custom_boundaries: Dict[str, Tuple[Number, Number]] = None
    ):
        lbls = [str(l).lower() for l in self.labels]
        if custom_boundaries is None:
            custom_boundaries = {}
        # FairRanges will automatically join defaults with custom boundaries
        bnd = FairRanges().load_fair_ranges(custom_boundaries)
        # Mismatched keys may lead to errant belief that a measure is within
        # the fair range when actually there was a mistake (eg. key was
        # misspelled)
        boundaries = {k: v for k, v in bnd.items() if k.lower().strip() in lbls}
        valid.validate_fair_boundaries(boundaries, lbls)
        self.boundaries = boundaries

    def __set_df(self, df: pd.DataFrame):
        if isinstance(df, pd.DataFrame):
            self.df = df.copy()
        else:
            err = "df must be a pandas DataFrame or Styler"
            try:
                if isinstance(df, pd.io.formats.style.Styler):
                    self.df = df.data.copy()
                else:
                    raise ValidationError(err)
            except:
                raise ValidationError(err)

    def __set_labels(self):
        """ Determines the locations of the strings containing measure names
            (within df); then reutrns a list of those measure names along
            with their location (one of ["columns", "index"]).
        """
        if self.df is None:
            pass
        else:
            if isinstance(self.df.index, pd.MultiIndex):
                if "Measure" in self.df.index.names:
                    label_type = "index"
                    labels = self.df.index.get_level_values("Measure")
                else:
                    label_type = "columns"
                    labels = self.df.columns.tolist()
            else:
                label_type = "columns"
                labels = self.df.columns.tolist()

            self.label_type, self.labels = label_type, labels

